src/pages/studyView/studyPageHeader/rightPanel/RightPanel.tsx

```diff
@@ -53,16 +57,88 @@ export default class RightPanel extends React.Component<IRightPanelProps, {}> {
         );
     }
 
+    @computed get isQueryLimitReached(): boolean {
+        if (this.props.store.selectedSamples.isComplete) {
+            return (
+                this.props.store.geneQueries.length *
+                    this.props.store.selectedSamples.result.length >
+                AppConfig.serverConfig.query_product_limit
+            );
+        } else {
+            return false;
+        }
+    }
+
+    @computed get geneLimit(): number {
+        if (this.props.store.selectedSamples.isComplete) {
+            return Math.floor(
+                AppConfig.serverConfig.query_product_limit /
+                    this.props.store.selectedSamples.result.length
+            );
+        } else {
+            // won't be used here
+            return 0;
+        }
+    }
+
+    @computed get oqlMessages() {
+        return getOqlMessages(this.props.store.geneQueries);
+    }
+
+    @computed get oqlSubmitError() {
+        if (this.props.store.isSingleNonVirtualStudyQueried) {
```

Comment:
what happens if it's multiple studies?

Reply:
that case is captured in the ensuing if blocks